### ISSUE TITLE
Incorrect report of needless collect  with the sequence from_fn().collect() follow by vec.into_iter().rev().collect()

### ISSUE THREAD

**Issue by galop1n** [rust-lang/rust-clippy#8132]
### Summary

The iterator in front of the first `collect` is not bi-directional. clippy fail at identify that the subsequent `rev()` would not accept it if the operation were chained as suggested.

### Lint Name

needless_collect

### Reproducer

I tried this code:

```rust
fn main() {
    let mut i = 0;
    let a = std::iter::from_fn(move || {
        i += 1;
        if i < 10 {
            Some(i)
        } else {
            None
        }
    })
    .collect::<Vec<usize>>();
    println!("{:?}", a.into_iter().rev().collect::<Vec<usize>>());
}
```

I saw this happen:

```
warning: avoid using `collect()` when not needed
  --> src/main.rs:11:6
   |
11 |     .collect::<Vec<usize>>();
   |      ^^^^^^^
12 |     println!("{:?}", a.into_iter().rev().collect::<Vec<usize>>());
   |                      ------------- the iterator could be used here instead
   |
   = note: `#[warn(clippy::needless_collect)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
help: use the original Iterator instead of collecting it and then producing a new one
   |
3  ~     
4  ~     println!("{:?}", std::iter::from_fn(move || {
5  +         i += 1;
6  +         if i < 10 {
7  +             Some(i)
8  +         } else {
 ...

warning: `playground` (bin "playground") generated 1 warning
```

I expected to see this happen:

### Additional Labels

l-suggestion-causes-error

**Comment by Testare** [comment]
Same issue here

**Comment by kyp44** [comment]
Having the same issue in commit `34a576b2be14289d6b95928508e557dcf217594c` of my project here: [https://github.com/kyp44/advent-of-code](https://github.com/kyp44/advent-of-code).

It seems that this lint should probably not flag if `.rev()` is used and the original iterator does not implement `DoubleEndedIterator`.
